File: model-value/src/main/java/com/bnorm/barkeep/model/BookValue.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model;

import javax.annotation.Nullable;

import org.jetbrains.annotations.NotNull;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableSet;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

@AutoValue
public abstract class BookValue implements Book {

  public static Builder builder() {
    return new AutoValue_BookValue.Builder();
  }

  public static JsonAdapter<BookValue> jsonAdapter(Moshi moshi) {
    return new AutoValue_BookValue.MoshiJsonAdapter(moshi);
  }

  @Override
  @Nullable
  public abstract ImmutableSet<Recipe> getRecipes();

  @Override
  public int compareTo(@NotNull Book other) {
    return Book.Companion.getCOMPARATOR().compare(this, other);
  }

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract BookValue build();

    public abstract Builder setId(Long id);

    public abstract Builder setTitle(String title);

    public abstract Builder setDescription(String description);

    public abstract Builder setOwner(User owner);

    abstract ImmutableSet.Builder<Recipe> recipesBuilder();

    public Builder addRecipe(Recipe recipe) {
      recipesBuilder().add(recipe);
      return this;
    }
  }
}

File: model-value/src/main/kotlin/com/bnorm/barkeep/model/BookValue.kt
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson

data class BookSpecValue(override val title: String?,
                         override val description: String? = null,
                         override val owner: User? = null) : BookSpec

data class BookValue(override val id: Long = -1,
                     override val title: String?,
                     override val description: String? = null,
                     override val owner: User? = null,
                     override val recipes: Set<Recipe> = emptySet()) : Book

object BookValueAdapter {
  @ToJson fun toJson(book: BookSpec): BookSpecValue {
    return BookSpecValue(book.title, book.description, book.owner)
  }

  @FromJson fun fromJson(json: BookSpecValue): BookSpec {
    return json
  }

  @ToJson fun toJson(book: Book): BookValue {
    return BookValue(book.id, book.title, book.description, book.owner, book.recipes)
  }

  @FromJson fun fromJson(json: BookValue): Book {
    return json
  }
}
